### PR TITLE
Dem concurrency = context.num_cores

### PR DESCRIPTION
**Pull request by smathermather**
I am wondering if it makes sense to have this as the default -- maximize the smoothness of the DEM products and maximize core usage. I can't think of a downside that can't be overridden, but perhaps I am missing something important.

### CODE REVIEW

**Comment by smathermather** [comment]
Of course, this only makes sense if each scale takes about the same amount of time to run... .

**Comment by pierotofy** [comment]
If you set the default to the max number of cores available, you might be spending time computing unnecessary DEM layers. Each gapfill step uses a progressively bigger radius for interpolation. The bigger the radius the longer it takes to compute the layer. At a certain point there's no need to compute layers further as the interpolation fills the entire raster (for dense point clouds, I found that `3` is a good value).

In short, this could cripple performance and do unnecessary computation on machines with lots of cores.

If smoother DEMs is the goal, I think the way to go would be to bring https://github.com/jblindsay/fp_denoise into the pipeline.

**Comment by smathermather** [comment]
You saved me some testing tonight. Closing.

**Comment by smathermather** [comment]
> If smoother DEMs is the goal, I think the way to go would be to bring https://github.com/jblindsay/fp_denoise into the pipeline.

Although, it would help if the project compiled:
```
nim c -d:release fp_denoise.nim
Hint: used config file '/home/useruser/nim/nim-0.19.0/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: fp_denoise [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: bitops [Processing]
Hint: algorithm [Processing]
Hint: unicode [Processing]
Hint: endians [Processing]
Hint: os [Processing]
Hint: times [Processing]
Hint: options [Processing]
Hint: typetraits [Processing]
Hint: strformat [Processing]
Hint: macros [Processing]
Hint: posix [Processing]
Hint: ospaths [Processing]
Hint: raster [Processing]
Hint: byte_order_reader [Processing]
Hint: unittest [Processing]
Hint: streams [Processing]
Hint: sets [Processing]
Hint: hashes [Processing]
Hint: terminal [Processing]
Hint: colors [Processing]
Hint: tables [Processing]
Hint: termios [Processing]
Hint: byte_order_writer [Processing]
Hint: sequtils [Processing]
raster.nim(400, 24) Error: type mismatch: got <seq[float64], nil>
but expected one of:
proc `==`(x, y: pointer): bool
  first type mismatch at position: 1
  required type: pointer
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: int64): bool
  first type mismatch at position: 1
  required type: int64
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: char): bool
  first type mismatch at position: 1
  required type: char
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: int16): bool
  first type mismatch at position: 1
  required type: int16
  but expression 'self.values' is of type: seq[float64]
proc `==`[T: tuple |
    object](x, y: T): bool
  first type mismatch at position: 1
  required type: T: tuple or object
  but expression 'self.values' is of type: seq[float64]
proc `==`[T](x, y: ref T): bool
  first type mismatch at position: 1
  required type: ref T
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: cstring): bool
  first type mismatch at position: 1
  required type: cstring
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: int8): bool
  first type mismatch at position: 1
  required type: int8
  but expression 'self.values' is of type: seq[float64]
proc `==`(x: type(nil); y: string): bool
  first type mismatch at position: 1
  required type: nil
  but expression 'self.values' is of type: seq[float64]
proc `==`[T: proc](x, y: T): bool
  first type mismatch at position: 1
  required type: T: proc
  but expression 'self.values' is of type: seq[float64]
proc `==`[T](x, y: openArray[T]): bool
  first type mismatch at position: 2
  required type: openarray[T]
  but expression 'nil' is of type: nil
proc `==`(x, y: string): bool
  first type mismatch at position: 1
  required type: string
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: float): bool
  first type mismatch at position: 1
  required type: float
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: float32): bool
  first type mismatch at position: 1
  required type: float32
  but expression 'self.values' is of type: seq[float64]
proc `==`[T](x, y: set[T]): bool
  first type mismatch at position: 1
  required type: set[T]
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: int32): bool
  first type mismatch at position: 1
  required type: int32
  but expression 'self.values' is of type: seq[float64]
proc `==`[I, T](x, y: array[I, T]): bool
  first type mismatch at position: 1
  required type: array[I, T]
  but expression 'self.values' is of type: seq[float64]
proc `==`(x, y: int): bool
  first type mismatch at position: 1
  required type: int
  but expression 'self.values' is of type: seq[float64]
proc `==`[T: SomeUnsignedInt](x, y: T): bool
  first type mismatch at position: 1
  required type: T: SomeUnsignedInt
  but expression 'self.values' is of type: seq[float64]
proc `==`[Enum: enum](x, y: Enum): bool
  first type mismatch at position: 1
  required type: Enum: enum
  but expression 'self.values' is of type: seq[float64]
proc `==`[T](x, y: seq[T]): bool
  first type mismatch at position: 2
  required type: seq[T]
  but expression 'nil' is of type: nil
proc `==`(x, y: bool): bool
  first type mismatch at position: 1
  required type: bool
  but expression 'self.values' is of type: seq[float64]
proc `==`(x: string; y: type(nil)): bool
  first type mismatch at position: 1
  required type: string
  but expression 'self.values' is of type: seq[float64]
proc `==`[T](x, y: ptr T): bool
  first type mismatch at position: 1
  required type: ptr T
  but expression 'self.values' is of type: seq[float64]
proc `==`(err1, err2: OSErrorCode): bool
  first type mismatch at position: 1
  required type: OSErrorCode
  but expression 'self.values' is of type: seq[float64]

expression: self.values == nil
```

**Comment by pierotofy** [comment]
There's another version at https://personalpages.manchester.ac.uk/staff/neil.mitchell/mdenoise/

The nim version would probably need to be ported to another language.

**Comment by smathermather** [comment]
It looks like the fix is done. I'll test this next week.

**Comment by smathermather** [comment]
> There's another version at https://personalpages.manchester.ac.uk/staff/neil.mitchell/mdenoise/
> 
> The nim version would probably need to be ported to another language.

I was wondering about the introduction of a new language... . The good news about mdenoise is it also can be applied to point clouds... .

I'll also test the following next week:
* Test changes to SMVS gamma
* Test use of OpenMVE for point cloud
* Test use of COLMAP for point cloud

**Comment by smathermather** [comment]
Definitely a noise reducer, but we have other work to do to improve the coherence of the point cloud before worrying about this level of noise reduction (fp_denoise):
screen shot 2018-11-21 at 4 53 45 pm
screen shot 2018-11-21 at 4 53 53 pm

**Comment by smathermather** [comment]
Real issue is here: https://github.com/OpenDroneMap/ODM/issues/653
